### ISSUE TITLE
class fields are shadowing Sequelize's attribute getters & setters

### ISSUE THREAD

**Issue by waqastanoli** [RobinBuschmann/sequelize-typescript#1236]
# Issue

## Versions

- sequelize: ^6.17.0
- sequelize-typescript: ^2.1.3
- typescript: ^4.5.4
- fastify: ^3.0.0

## Issue type

- [X] bug report 

## Actual behavior

```
[App] (sequelize) Warning: Model "Users" is declaring public class fields for attribute(s): "firstName", "lastName", "email", "userName", "phone", "dob", "password", "role_id", "role_title".
[App] These class fields are shadowing Sequelize's attribute getters & setters.
[App] See https://sequelize.org/main/manual/model-basics.html#caveat-with-public-class-field
```

## Expected behavior

a clean CLI log without this warning, what am i doing wrong.

## Steps to reproduce

I created simple models with annotations and use their objects to save or retrieve data from the database

## Related code

model

```ts
import {Table, Column, Model, DataType, Unique, BelongsToMany} from 'sequelize-typescript';
//import bcrypt from 'fastify-bcrypt'

import Roles from './roles';
import userRoles from './userRoles';
@Table({
  timestamps: true,
  paranoid: true  
})

export default class Users extends Model {
  @Column(DataType.VIRTUAL)
  get object() {
    return this.constructor.name.toLowerCase();
  }    


  @Column({field: 'firstName',
    type: DataType.STRING, 
    allowNull: false , 
    validate: {
      notNull: {msg: 'firstName is required'}
    }
  })
  public firstName!: string;


  @Column({field: 'lastName',
    type: DataType.STRING, 
    allowNull: false , 
    validate: {
      notNull: {msg: 'lastName is required'}
    }
  })
  public lastName!: string;

  @Unique({
    name: 'email',
    msg: 'email already taken, Please try a different email'
  })
  @Column({field: 'email',
    type: DataType.STRING, 
    allowNull: false , 
    validate: {
      notNull: {msg: 'email is required'}
    }
  })
  public email!: string;

  @Unique({
    name: 'userName',
    msg: 'username already taken, Please try a different username'
  })
  @Column({field: 'userName',
    type: DataType.STRING, 
    allowNull: false, 
    validate: {
      notNull: {msg: 'userName is required'},      
    }
  })
  public userName!: string;

  @Unique({
    name: 'phone',
    msg: 'phone already taken, Please try a different phone'
  })
  @Column({field: 'phone',
    type: DataType.STRING, 
    allowNull: false , 
    validate: {
      notNull: {msg: 'phone is required'}
    }
  })
  public phone!: string;

  @Column({field: 'dob',
    type: DataType.STRING, 
    allowNull: false , 
    validate: {
      notNull: {msg: 'dob is required'}
    }
  })
  public dob!: string;


  @Column({field: 'password',
    type: DataType.STRING, 
    allowNull: false , 
    
    validate: {
      notNull: {msg: 'password is required'}
    }
  })  
  public password?: string

  @Column({field: 'role_id',
    type: DataType.NUMBER, 
    allowNull: false , 
    
    validate: {
      notNull: {msg: 'role_id is required'}
    }
  })  
  public role_id?: number

  @Column({field: 'role_title',
    type: DataType.STRING, 
    allowNull: false , 
    
    validate: {
      notNull: {msg: 'role_title is required'}
    }
  })  
  public role_title?: string
  
  


  @BelongsToMany(() => Roles, {
    through: {
      model: () => userRoles,
      unique: false,
    },
    foreignKey: 'user_id',
    constraints: false,
  })
  roles: Roles[] | undefined;


  

  static encrypt(password, bcrypt) {  
      
    
    //password = bcrypt.hash(password, 10);
    return bcrypt.hash(password)
  .then(hash => {return hash})
  .catch(err => err.message)
  }

 

  toJSON() {
    return {...super.toJSON(), password: null};
  }
  
}
```
handler

```ts
const role_inserted = await Users.findByPk(id,{include: ["roles"]})
```
Although I am not facing any blocking errors right now, but this warning keep throwing and I am concerned if this will be a problem as my code base increases to accomodate more complex scenarios.
```
[App] (sequelize) Warning: Model "Users" is declaring public class fields for attribute(s): "firstName", "lastName", "email", "userName", "phone", "dob", "password", "role_id", "role_title".
[App] These class fields are shadowing Sequelize's attribute getters & setters.
[App] See https://sequelize.org/main/manual/model-basics.html#caveat-with-public-class-fields
```

**Comment by MonstraG** [comment]
For some reason, this started happening to me after I've updated nextjs from "12.1.0" to "12.1.1".

It also breaks completely, with all properties becoming `undefined` when attempting to read the model from db.

Enabling\disabling `swcMinify` doesn't help.

**Comment by waqastanoli** [comment]
@RobinBuschmann can you please guide us where we are going wrong, would appreciate your input on this.

**Comment by ephys** [comment]
Hey 👋🏼 
I'm the one that added this warning, it's actually thrown by Sequelize itself, it's not Sequelize-TypeScript

Unless the detection is a false positive, seeing this message typically means you're going to encounter issues.

Sequelize was designed long before public class fields were a thing, and the way Sequelize handles attributes is incompatible with public class fields.
See https://sequelize.org/docs/v6/core-concepts/model-basics/#caveat-with-public-class-fields
This is what we're doing to fix it: https://github.com/sequelize/sequelize/issues/14300

It only works if your TypeScript compiler strips these class fields out during compilation.
If you're using `tsc`, this can be done by writing your fields like this:

```typescript
 declare email: string;
```

instead of this:

```typescript
 public email!: string;
```

**Comment by waqastanoli** [comment]
> ```ts
> declare 
> ```

Thank you so much, this worked for me. Appreciate your contribution and helping out from bottom of my heart.
```ts
@Column({field: 'firstName',
    type: DataType.STRING, 
    allowNull: false , 
    validate: {
      notNull: {msg: 'firstName is required'}
    }
  })
  declare firstName: string;
```

 I have one curious question if any one else comes along same scenario.
 Can you give us a brief idea if some one ignores these warning, what kind of issues they might encounter?

**Comment by ephys** [comment]
Typically what is going to happen is that the value of field will be `undefined`

`Model.getDataValue` and `Model.get` will still return the right value though

**Comment by puchm** [comment]
Hey,

does anyone know how to use this `declare` syntax with Babel properly?
This is my babel config:
```json
{
    "presets": [
        "@babel/preset-typescript",
        "@babel/preset-env"
    ],
    "plugins": [
        ["@babel/plugin-transform-typescript", {
            "allowDeclareFields": true
        }],
        ["@babel/plugin-proposal-decorators", {
            "legacy": true
        }],
        ["@babel/plugin-proposal-class-properties", { "loose": true }],
        ["@babel/plugin-proposal-private-methods", { "loose": true }],
        ["@babel/plugin-proposal-private-property-in-object", { "loose": true }],
        "babel-plugin-transform-typescript-metadata"
    ]
}
```

This is my `tsconfig.json`:
```json
{
    "compilerOptions": {
        "target": "es2017",
        "lib": ["es2021"],
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "module": "ESNext",
        "moduleResolution": "node",
        "allowSyntheticDefaultImports": true
    }
}
```

In my code I can use the `declare` syntax without Babel failing to transpile. However, I can't access the properties so I guess the fields are still shadowing sequelize's fields. I am also still getting the warning about this.

Any help would be greatly appreciated!

**Comment by ephys** [comment]
Babel opted to ignore `declare` if the field is decorated, see https://github.com/RobinBuschmann/sequelize-typescript/issues/778#issuecomment-1007615424 for a workaround

**Comment by max-lt** [comment]
Got the same "problem" when I changed "target": "ES2017", to "target": "ES2022" / "ESNext" in typescript compilerOptions.
I wasn't using nest Decorators.
I used the `declare` solution that is logical anyway as those attributes are virtual, I think it's better as is.

**Comment by heblol** [comment]
Hi thanks, changing the variable to `declare VARIABLE_NAME: TYPE` worked for me. However, I now get a typescript error `Decorators are not valid here.ts` at the location of the `@` sign.

Example code:
```
import {
  Model,
  Table,
  Column,
  DataType,
  Index,
  Sequelize,
  ForeignKey,
} from "sequelize-typescript";

export interface groupAttributes {
  id?: string;
  name?: string;
  created_at?: Date;
  updated_at?: Date;
  invite_uuid?: string;
  invite_open?: boolean;
  beta?: boolean;
  default_status?: string;
  origin?: string;
  email?: string;
  password?: string;
  pincode?: number;
  login_name?: string;
  address?: string;
  created_at_eetlijst?: Date;
  created_at_ip?: string;
  city?: string;
  description?: string;
  active?: boolean;
  default_close_time?: Date;
  created_by?: string;
  tablet_id?: string;
}

@Table({
  tableName: "group",
  timestamps: false,
})
export class group
  extends Model<groupAttributes, groupAttributes>
  implements groupAttributes
{
  @Column({                              // The error shows up in VSC at the `@` sign.
    primaryKey: true,
    type: DataType.UUID,
    defaultValue: Sequelize.literal("gen_random_uuid()"),
  })
  ...

``

**Comment by guilhermelim** [comment]
To resolve the warning about "class fields are shadowing Sequelize's attribute getters & setters," it's quite simple. The solution is to use the `declare` modifier for your model class fields. This prevents the issue where public fields might overshadow Sequelize's automatic getters and setters.

### Example Module Generating the Warning

Below is an example of a module that might trigger the mentioned warning:

```ts
console.warn
      (sequelize) Warning: Model "ProductModel" is declaring public class fields for attribute(s): "id", "purchasePrice", "description", "updatedAt", "createdAt", "stock", "name".
      These class fields are shadowing Sequelize's attribute getters & setters.
      See https://sequelize.org/main/manual/model-basics.html#caveat-with-public-class-fields
```
![image](https://github.com/user-attachments/assets/3782d195-d08e-4992-a21c-c73af316264c)

```ts
import { PrimaryKey, Column, Model, Table } from 'sequelize-typescript';

@Table({
  tableName: 'products',
  timestamps: false,
})
export default class ProductModel extends Model {
  @PrimaryKey
  @Column({ allowNull: false })
  id: string;

  @Column({ allowNull: false })
  purchasePrice: number;

  @Column({ allowNull: false })
  description: string;

  @Column({ allowNull: false })
  updatedAt: Date;

  @Column({ allowNull: false })
  createdAt: Date;

  @Column({ allowNull: false })
  stock: number;

  @Column({ allowNull: false })
  name: string;
}
```

### Example of the Module Fixing the Warning

To fix the issue, update the fields to use the `declare` modifier:

```ts
import { PrimaryKey, Column, Model, Table } from 'sequelize-typescript';

@Table({
  tableName: 'products',
  timestamps: false,
})
export default class ProductModel extends Model {
  @PrimaryKey
  @Column({ allowNull: false })
  declare id: string;

  @Column({ allowNull: false })
  declare purchasePrice: number;

  @Column({ allowNull: false })
  declare description: string;

  @Column({ allowNull: false })
  declare updatedAt: Date;

  @Column({ allowNull: false })
  declare createdAt: Date;

  @Column({ allowNull: false })
  declare stock: number;

  @Column({ allowNull: false })
  declare name: string;
}
```

### Explanation

By using the `declare` modifier, you indicate to TypeScript that the property is part of the class definition but without actually declaring a public field. This prevents the public fields in the class from interfering with Sequelize's automatic getters and setters, addressing the issue described in the [Sequelize documentation](https://sequelize.org/main/manual/model-basics.html#caveat-with-public-class-fields).

**Comment by David-Kluzak** [comment]
> Got the same "problem" when I changed "target": "ES2017", to "target": "ES2022" / "ESNext" in typescript compilerOptions. I wasn't using nest Decorators. I used the `declare` solution that is logical anyway as those attributes are virtual, I think it's better as is.

Yes, that's the cause in my case probably, too. My program worked and then after it was placed into a new tsconfig.json context with different compiler target, I stopped to work with the discussed issue.

**Comment by dejf-js** [comment]
https://github.com/sequelize/sequelize-typescript/issues/1236#issuecomment-1345237370
>Got the same "problem" when I changed "target": "ES2017", to "target": "ES2022" / "ESNext" in typescript compilerOptions. I wasn't using nest Decorators. I used the `declare` solution that is logical anyway as those attributes are virtual, I think it's better as is.

Yes. After moving my code to another tsconfig.js context having different compiler "target", my code suddenly stopped working.

**Comment by dejf-js** [comment]
I don't know if we need to use `declare` for relational attributes like for example:
``` javascript
  @BelongsTo(() => Activity)
  declare activity: Activity;
```
and other associations.
